oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs

```diff
@@ -548,64 +586,80 @@ impl ManifestStorage {
-                conn.exec(|conn| {
-                    diesel::delete(
-                        chunks::table.filter(chunks::chunk_id.eq_any(pending_chunk_ids_chunk)),
-                    )
-                    .execute(conn)
-                })?;
+                        .execute(conn)
+            }).await?;
+
+            for pending_chunk_id in chunk_ids
+                .chunks(LOCAL_DATABASE_MAX_VARIABLE_NUMBER)
+                .map(|chunk| chunk.to_vec())
+            {
+                self.conn
+                    .exec(move |conn| {
+                        diesel::delete(
+                            chunks::table.filter(chunks::chunk_id.eq_any(pending_chunk_id)),
+                        )
+                        .execute(conn)
+                    })
+                    .await?;
             }
             cache.pending_chunk_ids = None;
         }
         Ok(())
     }
 
-    pub fn flush_cache_ahead_of_persistance(&self) -> FSResult<()> {
+    pub async fn flush_cache_ahead_of_persistance(&self) -> FSResult<()> {
         let items = self
             .caches
             .lock()
             .expect("Mutex is poisoned")
             .iter()
             .map(|(entry_id, entry)| (*entry_id, entry.clone()))
-            .collect::<Vec<(EntryID, Arc<Mutex<CacheEntry>>)>>();
+            .collect::<Vec<(EntryID, Arc<AsyncMutex<CacheEntry>>)>>();
         for (entry_id, lock) in items {
-            self.ensure_manifest_persistent_lock(entry_id, lock)?;
+            self.ensure_manifest_persistent_lock(entry_id, lock).await?;
         }
         Ok(())
     }
 
     /// This method is not used in the code base but it is still tested
     /// as it might come handy in a cleanup routine later
     #[deprecated]
-    pub fn clear_manifest(&self, entry_id: EntryID) -> FSResult<()> {
+    pub async fn clear_manifest(&self, entry_id: EntryID) -> FSResult<()> {
         // Remove from cache
         let cache_entry = self.get_or_insert_default_cache_entry(entry_id);
-        let mut cache_unlock = cache_entry.lock().expect("Mutex is poisoned");
+        let mut cache_unlock = cache_entry.lock().await;
 
         // Remove from local database
-        let mut conn = self.conn.lock_connection();
+        // TODO: do we need to lock the database for the operation ?
+        // let mut conn = self.conn.lock_connection();
 
-        let deleted = conn.exec(|conn| {
-            diesel::delete(vlobs::table.filter(vlobs::vlob_id.eq((*entry_id).as_ref())))
-                .execute(conn)
-        })? > 0;
+        let deleted = self
+            .conn
+            .exec(move |conn| {
+                let vlob_id = (*entry_id).as_ref();
+                diesel::delete(vlobs::table.filter(vlobs::vlob_id.eq(vlob_id))).execute(conn)
+            })
+            .await?
+            > 0;
 
-        if let Some(pending_chunk_ids) = &cache_unlock.pending_chunk_ids {
+        if let Some(pending_chunk_ids) = cache_unlock.pending_chunk_ids.clone() {
```

Comment:
Why clone ?

Reply:
Lifetime concern when sending the value across thread

Reply:
Which thread ?

Reply:
take a look to `platform_local_db`